### PR TITLE
Support max_rows, max_cols in gr.Dataframe

### PR DESCRIPTION
**Pull request by 1lint**
# Description
Add support for `max_rows `and `max_cols `parameters in `gr.Dataframe` by slicing input dataframes in the `postprocess` function. 

These parameters didn't seem to do anything when I tried to use them, then I saw it was an open issue

Here's a quick UI test to see that it works
```
import gradio as gr
import pandas as pd
import numpy as np

with gr.Blocks() as demo:
    df_display = gr.Dataframe(max_rows = 2, max_cols = 4)

    examples_rows = 5
    example_cols = 10

    with gr.Row():
        gr.Examples(
            inputs = df_display,
            examples=[
                np.zeros((examples_rows, example_cols)), # numpy array
                [[0]*example_cols]*examples_rows, # python array
                pd.DataFrame([[0]*example_cols]*examples_rows), # pandas array
                {
                    "headers" : list(range(example_cols)), # dict array
                    "data" : [[0]*example_cols]*examples_rows
                }
            ],
            examples_per_page=5,
        )

demo.launch()
```

Closes: # (issue)
https://github.com/gradio-app/gradio/issues/2209
# Checklist:

- [x ] I have performed a self-review of my own code
- [x ] I have added a short summary of my change to the CHANGELOG.md
- [x ] My code follows the style guidelines of this project
- [x ] I have commented my code in hard-to-understand areas
- [x ] I have made corresponding changes to the documentation
- [x ] I have added tests that prove my fix is effective or that my feature works
- [x ] New and existing unit tests pass locally with my changes

# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

### CODE REVIEW

**Comment by pngwn** [comment]
I think we should probably hold off on adding features with 4.0 around the corner. We may want to revisit some of the Dataframe features and kwargs. 

Cc @abidlabs

**Comment by abidlabs** [comment]
@pngwn while it's true that we shouldn't add new features to the `gr.Dataframe`, this is technically a bug fix since these parameters are already in the constructor of `gr.Dataframe`. 

Imo we should review this PR and merge in if the fix looks good because if @1lint hadn't opened this PR, @hannahblair or I would have taken this issue on as part of 4.0 work anyways to ensure that the `gr.Dataframe` is in a good state before the 4.0 release.

**Comment by pngwn** [comment]
Sounds good.

**Comment by 1lint** [comment]
Great glad I could help! I just pushed an updated test to cover string inputs, and a related bug fix for my 1st commit

File: gradio/components.py [hunk 1]
```diff
@@ -3051,16 +3051,22 @@ def postprocess(
...
+            dataframe = pd.read_csv(y, nrows=self.max_rows)
+            dataframe = dataframe.loc[:, dataframe.columns[: self.max_cols]]
             return {
                 "headers": list(dataframe.columns),
                 "data": Dataframe.__process_markdown(
                     dataframe.to_dict(orient="split")["data"], self.datatype
                 ),
             }
         if isinstance(y, pd.DataFrame):
+            y = y.loc[: self.max_rows - 1, y.columns[: self.max_cols]]
```
Comment (freddyaboulton) [inline gradio/components.py:3069]: We should be using .iloc here for the rows as .loc is label-based and will not work if the dataframe has a custom index.

image

File: gradio/components.py [hunk 2]
```diff
@@ -3051,16 +3051,22 @@ def postprocess(
...
         if isinstance(y, dict):
+            y["headers"] = y["headers"][: self.max_cols]
+            y["data"] = list(
+                map(lambda row: row[: self.max_cols], y["data"][: self.max_rows])
+            )
             return y
         if isinstance(y, str):
-            dataframe = pd.read_csv(y)
+            dataframe = pd.read_csv(y, nrows=self.max_rows)
+            dataframe = dataframe.loc[:, dataframe.columns[: self.max_cols]]
```
Comment (freddyaboulton) [inline gradio/components.py:3061]: .loc is ok here since we're only indexing the columns by name

**Review by freddyaboulton — COMMENTED** [review]
Thank you for the contribution @1lint ! And for writing the test.

Broader question for @pngwn and @abidlabs - shouldn't this be a front-end change though? I think it would be best if these parameters worked in conjunction with pagination to control the data that's initially displayed. Doing it in the backend means the data outside of `max_rows` and `max_cols` is "lost".

File: gradio/components.py (hunk 1 above)
Comment (1lint) [reply gradio/components.py:3069]: Good catch! I will update the slicing and test accordingly

File: gradio/components.py (hunk 2 above)
Comment (1lint) [reply gradio/components.py:3061]: Thanks for pointing this out, will update it to iloc as well for consistency, should be more efficient too

**Comment by 1lint** [comment]
@freddyaboulton thanks for the great catch on `loc` vs `iloc` usage in the dataframe!

That makes sense that the user may expect the `max_rows/max_cols` parameters to only affect the dataframe display, but preserve the rest of the data when using the dataframe as an input component. 
I would be worried that the user may be using these parameters because they have a large dataframe, in which case the truncated data may take up alot of bandwidth during the round trip to/from the frontend. 

It may be better to keep the truncated data in some sort of internal state component, and only send to the frontend the data that the user can see/modify. Then the (potentially) modified data read from the frontend can be recombined with the full data saved in the state component

This is essentially what I'm doing for an app to apply SQL queries to HF datasets - the user can select what dataset to load, but only the first 40 rows of the dataset are shown on the frontend, while the entire dataset is saved in gr.State. The user can quickly try out sql queries on this first 40 rows to see how it affects the output inside gr.Dataframe. Finally when they are happy with the result, they can apply the query to the whole dataset in gr.State and download the result.

**Comment by pngwn** [comment]
This is why I wanted to leave this for now. It really depends on how people will use it but because we have sortable data frames the expectation would be that the first n rows would be returned of the complete sorted dataset. Likewise, if someone is aggregating the data in other output components then those aggregations should operate on the entire dataset, not just a subset of it.

**Comment by pngwn** [comment]
I think some of these cases will require significant changes to work correctly and there may need to be breaking changes too. Right now whatever the user function returns is what gets displayed, whatever gets passed back is the data you can operate on. Internal state would break this expectation.  

Adding in some concept of internal state, would probably cause issues for the API as well.

**Comment by abidlabs** [comment]
I agree that `max_rows` and `max_cols` should filter on the frontend. If only a subset of data should be returned from the backend, that can be a part of the prediction function.

@1lint let's go ahead and close this PR since we'd need to approach quite differently. And @freddyaboulton let's move #2209 to the 4.0 milestone?

**Comment by 1lint** [comment]
Sounds good @abidlabs, good point this approach should be handled in the prediction function

**Pull request closed without merging**